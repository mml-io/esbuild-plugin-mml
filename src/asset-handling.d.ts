import * as esbuild from "esbuild";
import { Asset } from "./types";
export declare function createAssetHandler(sourceRoot: string, outdir: string, assetDir: string, assetPrefix: string, assets: Asset[], build: esbuild.PluginBuild): (args: esbuild.OnLoadArgs) => Promise<esbuild.OnLoadResult | null>;
