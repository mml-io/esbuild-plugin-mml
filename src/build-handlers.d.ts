import * as esbuild from "esbuild";
import { DiscoveredEntryPoints } from "./types";
export declare function createEntryPointResolver(discoveredEntryPoints: DiscoveredEntryPoints, build: esbuild.PluginBuild): (args: esbuild.OnResolveArgs) => Promise<esbuild.OnResolveResult | null>;
export declare function createMMLImportResolver(build: esbuild.PluginBuild): (args: esbuild.OnResolveArgs) => Promise<esbuild.OnResolveResult>;
export declare function createMMLRefHandler(sourceRoot: string, entryPointToOutputUrl: Record<string, string>, documentPrefix: string): (args: esbuild.OnLoadArgs) => esbuild.OnLoadResult;
export declare function createMMLTSXRefHandler(sourceRoot: string, entryPointToOutputUrl: Record<string, string>, documentPrefix: string): (args: esbuild.OnLoadArgs) => esbuild.OnLoadResult;
export declare function createTSXHandler(): (args: esbuild.OnLoadArgs) => Promise<esbuild.OnLoadResult>;
export declare function createWorldHandler(): (args: esbuild.OnLoadArgs) => Promise<esbuild.OnLoadResult>;
export declare function createHTMLHandler(): (args: esbuild.OnLoadArgs) => Promise<esbuild.OnLoadResult>;
