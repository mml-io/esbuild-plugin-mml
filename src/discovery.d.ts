import * as esbuild from "esbuild";
import { DiscoveredEntryPoints } from "./types";
export declare function doDiscoveryBuild(esbuildOptions: esbuild.BuildOptions): Promise<DiscoveredEntryPoints>;
