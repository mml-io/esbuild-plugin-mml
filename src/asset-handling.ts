import * as esbuild from "esbuild";
import * as path from "path";
import fsp from "node:fs/promises";
import { Asset } from "./types";

export function createAssetHandler(
  sourceRoot: string,
  outdir: string,
  assetDir: string,
  assetPrefix: string,
  assets: Asset[],
  build: esbuild.PluginBuild,
) {
  return async (
    args: esbuild.OnLoadArgs,
  ): Promise<esbuild.OnLoadResult | null> => {
    const extension = path.extname(args.path);

    // Check if this file should be handled as an asset
    const loader = build.initialOptions.loader?.[extension];
    if (loader !== "file") {
      return null;
    }

    // Generate asset file name based on the source path
    const relativePath = path.relative(sourceRoot, args.path);
    const fileName = relativePath.replace(/[/\\]/g, "_");

    // Build full asset URL
    const assetUrl = assetPrefix + fileName;

    // Determine output path
    const outputPath = assetDir
      ? path.join(outdir, assetDir, fileName)
      : path.join(outdir, fileName);

    // Copy the asset to the output directory
    await fsp.mkdir(path.dirname(outputPath), { recursive: true });
    await fsp.copyFile(args.path, outputPath);

    // Track the asset
    assets.push({
      fileName,
      url: assetUrl,
      path: outputPath,
    });

    return {
      contents: assetUrl,
      loader: "text",
    };
  };
}
