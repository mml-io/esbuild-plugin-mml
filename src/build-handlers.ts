import * as esbuild from "esbuild";
import * as path from "path";
import fsp from "node:fs/promises";
import { DiscoveredEntryPoints } from "./types";

export function createEntryPointResolver(
  discoveredEntryPoints: DiscoveredEntryPoints,
  build: esbuild.PluginBuild,
) {
  return async (
    args: esbuild.OnResolveArgs,
  ): Promise<esbuild.OnResolveResult | null> => {
    if (args.kind !== "entry-point" || args.pluginData === "mml-skip-resolve") {
      return null;
    }

    const resolvedPath = await build.resolve(args.path, {
      resolveDir: args.resolveDir,
      kind: args.kind,
      pluginData: "mml-skip-resolve",
    });

    const namespace = discoveredEntryPoints.tsx.has(resolvedPath.path)
      ? "mml-build-tsx"
      : discoveredEntryPoints.html.has(resolvedPath.path)
        ? "mml-build-html"
        : discoveredEntryPoints.world.has(resolvedPath.path)
          ? "mml-build-world"
          : null;

    return namespace
      ? {
          path: resolvedPath.path,
          namespace,
          pluginData: resolvedPath.pluginData as unknown,
        }
      : null;
  };
}

export function createMMLImportResolver(build: esbuild.PluginBuild) {
  return async (
    args: esbuild.OnResolveArgs,
  ): Promise<esbuild.OnResolveResult> => {
    const importPath = args.path.replace(/^mml:/, "");
    const resolvedPath = await build.resolve(importPath, {
      resolveDir: args.resolveDir,
      kind: args.kind,
      pluginData: "mml-skip-resolve",
    });

    if (resolvedPath.errors.length > 0) {
      return {
        errors: resolvedPath.errors.map((error) => ({
          ...error,
          text: `Failed to resolve ${importPath} from ${args.importer}. ${error.text}`,
        })),
      };
    }

    return {
      path: resolvedPath.path,
      namespace: resolvedPath.path.endsWith(".html")
        ? "mml-build-ref"
        : "mml-build-tsx-ref",
      pluginData: resolvedPath.pluginData as string,
    };
  };
}

function createRefHandler(
  sourceRoot: string,
  entryPointToOutputUrl: Record<string, string>,
  documentPrefix: string,
  isTSX = false,
) {
  return (args: esbuild.OnLoadArgs): esbuild.OnLoadResult => {
    const relativePath = path.relative(sourceRoot, args.path);
    let outputUrl = entryPointToOutputUrl[relativePath];

    if (!outputUrl) {
      const baseName = path.basename(relativePath, path.extname(relativePath));
      const dirPath = path.dirname(relativePath);
      const pathParts = [dirPath, baseName].filter(
        (part) => part && part !== ".",
      );

      outputUrl =
        pathParts
          .join("-")
          .replace(isTSX ? /[^a-zA-Z0-9\-_]/g : /[/\\]/g, "-") + ".html";

      if (!isTSX) {
        entryPointToOutputUrl[relativePath] = outputUrl;
      }
    }

    return {
      contents: documentPrefix + outputUrl,
      loader: "text",
      resolveDir: path.dirname(args.path),
      watchFiles: [args.path],
    };
  };
}

export function createMMLRefHandler(
  sourceRoot: string,
  entryPointToOutputUrl: Record<string, string>,
  documentPrefix: string,
) {
  return createRefHandler(
    sourceRoot,
    entryPointToOutputUrl,
    documentPrefix,
    false,
  );
}

export function createMMLTSXRefHandler(
  sourceRoot: string,
  entryPointToOutputUrl: Record<string, string>,
  documentPrefix: string,
) {
  return createRefHandler(
    sourceRoot,
    entryPointToOutputUrl,
    documentPrefix,
    true,
  );
}

async function loadFileContent(
  args: esbuild.OnLoadArgs,
  loader: "tsx" | "text",
) {
  const content = await fsp.readFile(args.path, "utf-8");
  return {
    contents: content,
    loader,
    resolveDir: path.dirname(args.path),
    watchFiles: [args.path],
  };
}

export const createTSXHandler = () => (args: esbuild.OnLoadArgs) =>
  loadFileContent(args, "tsx");

export const createWorldHandler = () => (args: esbuild.OnLoadArgs) =>
  loadFileContent(args, "tsx");

export const createHTMLHandler = () => (args: esbuild.OnLoadArgs) =>
  loadFileContent(args, "text");
