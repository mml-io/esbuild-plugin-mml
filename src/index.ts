export * from "./plugin";
export * from "./world-config";
