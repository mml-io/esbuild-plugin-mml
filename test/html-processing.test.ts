import * as esbuild from "esbuild";
import { mml } from "../src";
import fsp from "node:fs/promises";
import path from "node:path";
import { waitForDispose, walk } from "./test-helpers";

describe("HTML Processing", () => {
  const testOutdir = "test/generated/html-processing";

  beforeEach(async () => {
    await fsp.rm(testOutdir, { recursive: true, force: true });
  });

  it("should handle HTML files as entry points", async () => {
    const outdir = path.join(testOutdir, "html-entry-points");
    await fsp.rm(outdir, { recursive: true, force: true });
    
    const config = {
      outdir,
      entryPoints: ["mml:test/src/c/d.html"],
      loader: {
        ".glb": "file" as const,
      },
      plugins: [mml({
        assetDir: "assets",
      })],
    };

    const result = await esbuild.build(config);

    // Check for build errors and fail the test if any exist
    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    for await (const { path, content } of walk(outdir)) {
      expect(content).toMatchSnapshot(path);
    }
  });

  it("should handle mml: imports of HTML files", async () => {
    const outdir = path.join(testOutdir, "mml-html-imports");
    await fsp.rm(outdir, { recursive: true, force: true });
    
    // Create a test file that imports an HTML file using mml: prefix
    const testFile = path.join("test/src", "html-importer.ts");
    await fsp.writeFile(testFile, `
      import htmlContent from "mml:./c/d.html";
      
      export default function() {
        return htmlContent;
      }
    `);

    try {
      const config = {
        outdir,
        entryPoints: [`mml:test/src/html-importer.ts`],
        loader: {
          ".glb": "file" as const,
        },
        plugins: [mml({
          assetDir: "assets",
        })],
      };

      const result = await esbuild.build(config);

      // Check for build errors
      if (result.errors.length > 0) {
        throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
      }

      await waitForDispose();

      // Verify the output files exist
      const manifestPath = path.join(outdir, "manifest.json");
      const manifestExists = await fsp.access(manifestPath).then(() => true).catch(() => false);
      expect(manifestExists).toBe(true);

      for await (const { path, content } of walk(outdir)) {
        expect(content).toMatchSnapshot(path);
      }
    } finally {
      // Clean up the test file
      await fsp.rm(testFile, { force: true });
    }
  });

  it("should process the existing c/d.html file", async () => {
    const outdir = path.join(testOutdir, "subdirectory-html");
    await fsp.rm(outdir, { recursive: true, force: true });
    
    const config = {
      outdir,
      entryPoints: ["mml:test/src/c/d.html"],
      loader: {
        ".glb": "file" as const,
      },
      plugins: [mml({
        assetDir: "assets",
      })],
    };

    const result = await esbuild.build(config);

    // Check for build errors and fail the test if any exist
    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    for await (const { path, content } of walk(outdir)) {
      expect(content).toMatchSnapshot(path);
    }
  });

  it("should handle HTML files with various content types", async () => {
    const outdir = path.join(testOutdir, "content-structures");
    
    // Create test HTML files with different structures
    await fsp.mkdir(path.join(testOutdir, "html-test"), { recursive: true });
    
    const simpleHtml = `<!DOCTYPE html><html><head><title>Simple</title></head><body><h1>Hello</h1></body></html>`;
    const complexHtml = `
      <!DOCTYPE html>
      <html lang="en">
        <head>
          <meta charset="UTF-8">
          <meta name="viewport" content="width=device-width, initial-scale=1.0">
          <title>Complex HTML</title>
        </head>
        <body>
          <header>
            <nav>
              <ul>
                <li><a href="#home">Home</a></li>
                <li><a href="#about">About</a></li>
              </ul>
            </nav>
          </header>
          <main>
            <section id="content">
              <article>
                <h1>Article Title</h1>
                <p>Some content here.</p>
              </article>
            </section>
          </main>
          <footer>
            <p>&copy; 2023 Test</p>
          </footer>
        </body>
      </html>
    `;
    
    await fsp.writeFile(path.join(testOutdir, "html-test", "simple.html"), simpleHtml);
    await fsp.writeFile(path.join(testOutdir, "html-test", "complex.html"), complexHtml);
    
    const config = {
      outdir,
      entryPoints: [
        `mml:${testOutdir}/html-test/simple.html`,
        `mml:${testOutdir}/html-test/complex.html`
      ],
      plugins: [mml()],
    };

    const result = await esbuild.build(config);

    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    // Check that both HTML files were processed
    const manifestPath = path.join(outdir, "manifest.json");
    const manifest = JSON.parse(await fsp.readFile(manifestPath, "utf-8"));
    
    expect(Object.keys(manifest.documentNameToPath)).toHaveLength(2);
    expect(Object.keys(manifest.documentNameToPath)).toContain("test-generated-html-processing-html-test-simple.html");
    expect(Object.keys(manifest.documentNameToPath)).toContain("test-generated-html-processing-html-test-complex.html");

    for await (const { path, content } of walk(outdir)) {
      expect(content).toMatchSnapshot(path);
    }
  });

  it("should handle HTML files with embedded scripts and styles", async () => {
    const outdir = path.join(testOutdir, "embedded-content");
    
    // Create HTML with embedded scripts and styles
    await fsp.mkdir(path.join(testOutdir, "embedded-test"), { recursive: true });
    
    const embeddedHtml = `
      <!DOCTYPE html>
      <html>
        <head>
          <title>Embedded Content</title>
          <style>
            body { background-color: #f0f0f0; }
            .container { max-width: 800px; margin: 0 auto; }
            h1 { color: #333; }
          </style>
        </head>
        <body>
          <div class="container">
            <h1 id="title">Dynamic Title</h1>
            <p>This page has embedded JavaScript and CSS.</p>
            <button onclick="changeTitle()">Change Title</button>
          </div>
          
          <script>
            function changeTitle() {
              document.getElementById('title').textContent = 'Title Changed!';
            }
            
            // Initialize the page
            document.addEventListener('DOMContentLoaded', function() {
              console.log('Page loaded with embedded scripts');
            });
          </script>
        </body>
      </html>
    `;
    
    await fsp.writeFile(path.join(testOutdir, "embedded-test", "embedded.html"), embeddedHtml);
    
    const config = {
      outdir,
      entryPoints: [`mml:${testOutdir}/embedded-test/embedded.html`],
      plugins: [mml()],
    };

    const result = await esbuild.build(config);

    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    // Verify the HTML file was processed and the content is preserved
    const outputHtml = path.join(outdir, "test-generated-html-processing-embedded-test-embedded.html");
    const content = await fsp.readFile(outputHtml, "utf-8");
    
    expect(content).toContain("<style>");
    expect(content).toContain("background-color: #f0f0f0");
    expect(content).toContain("<script>");
    expect(content).toContain("function changeTitle()");

    for await (const { path, content } of walk(outdir)) {
      expect(content).toMatchSnapshot(path);
    }
  });

  it("should handle complex HTML structures and nested elements", async () => {
    const outdir = path.join(testOutdir, "complex-structures");
    
    // Create HTML with deeply nested and complex structures
    await fsp.mkdir(path.join(testOutdir, "complex-test"), { recursive: true });
    
    const complexStructureHtml = `
      <!DOCTYPE html>
      <html>
        <head>
          <title>Complex Structure</title>
        </head>
        <body>
          <div class="wrapper">
            <header class="main-header">
              <nav class="navigation">
                <ul class="nav-list">
                  <li class="nav-item">
                    <a href="#section1" class="nav-link">
                      <span class="link-text">Section 1</span>
                      <span class="link-icon">â†’</span>
                    </a>
                    <ul class="sub-nav">
                      <li><a href="#subsection1">Sub 1</a></li>
                      <li><a href="#subsection2">Sub 2</a></li>
                    </ul>
                  </li>
                </ul>
              </nav>
            </header>
            
            <main class="content">
              <section id="section1" class="content-section">
                <article class="article">
                  <header class="article-header">
                    <h1 class="article-title">Complex Article</h1>
                    <div class="article-meta">
                      <time datetime="2023-01-01">January 1, 2023</time>
                      <address class="author">
                        <a rel="author" href="/author">John Doe</a>
                      </address>
                    </div>
                  </header>
                  
                  <div class="article-content">
                    <p>This is a paragraph with <strong>bold</strong> and <em>italic</em> text.</p>
                    <blockquote cite="https://example.com">
                      <p>This is a quoted paragraph.</p>
                      <footer>
                        â€” <cite>Famous Author</cite>
                      </footer>
                    </blockquote>
                    
                    <figure class="image-figure">
                      <img src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAwIiBoZWlnaHQ9IjEwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cmVjdCB3aWR0aD0iMTAwIiBoZWlnaHQ9IjEwMCIgZmlsbD0iIzMzMzMzMyIvPjwvc3ZnPg==" alt="Test image" />
                      <figcaption>This is a test image caption</figcaption>
                    </figure>
                    
                    <details class="expandable-content">
                      <summary>Click to expand</summary>
                      <div class="hidden-content">
                        <p>This content was hidden.</p>
                        <ul>
                          <li>Item 1</li>
                          <li>Item 2</li>
                        </ul>
                      </div>
                    </details>
                  </div>
                </article>
              </section>
            </main>
            
            <aside class="sidebar">
              <section class="widget">
                <h3>Related Links</h3>
                <nav class="widget-nav">
                  <ul>
                    <li><a href="#related1">Related 1</a></li>
                    <li><a href="#related2">Related 2</a></li>
                  </ul>
                </nav>
              </section>
            </aside>
            
            <footer class="main-footer">
              <div class="footer-content">
                <p>&copy; 2023 Complex Structure Test</p>
              </div>
            </footer>
          </div>
        </body>
      </html>
    `;
    
    await fsp.writeFile(path.join(testOutdir, "complex-test", "complex.html"), complexStructureHtml);
    
    const config = {
      outdir,
      entryPoints: [`mml:${testOutdir}/complex-test/complex.html`],
      plugins: [mml()],
    };

    const result = await esbuild.build(config);

    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    // Verify the complex HTML structure was preserved
    const outputHtml = path.join(outdir, "test-generated-html-processing-complex-test-complex.html");
    const content = await fsp.readFile(outputHtml, "utf-8");
    
    expect(content).toContain('<nav class="navigation">');
    expect(content).toContain('<ul class="sub-nav">');
    expect(content).toContain('<article class="article">');
    expect(content).toContain('<figure class="image-figure">');
    expect(content).toContain('<details class="expandable-content">');
    expect(content).toContain('<aside class="sidebar">');

    for await (const { path, content } of walk(outdir)) {
      expect(content).toMatchSnapshot(path);
    }
  });
}); 