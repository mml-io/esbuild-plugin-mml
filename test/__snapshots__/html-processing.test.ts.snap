// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML Processing should handle HTML files as entry points: generated/html-processing/html-entry-points/d.js 1`] = `
"(() => {
  var d_default = "<html>\\n  <body>\\n    <p>I'm html!</p>\\n  </body>\\n</html>\\n";
})();
"
`;

exports[`HTML Processing should handle HTML files as entry points: generated/html-processing/html-entry-points/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-src-c-d.html": "test-src-c-d.html"
  },
  "assetNameToPath": {},
  "documentPrefix": "ws:///",
  "assetPrefix": "/"
}"
`;

exports[`HTML Processing should handle HTML files as entry points: generated/html-processing/html-entry-points/test-src-c-d.html 1`] = `
"<html>
  <body>
    <p>I'm html!</p>
  </body>
</html>
"
`;

exports[`HTML Processing should handle HTML files with embedded scripts and styles: generated/html-processing/embedded-content/embedded.js 1`] = `
"(() => {
  var embedded_default = \`
      <!DOCTYPE html>
      <html>
        <head>
          <title>Embedded Content</title>
          <style>
            body { background-color: #f0f0f0; }
            .container { max-width: 800px; margin: 0 auto; }
            h1 { color: #333; }
          </style>
        </head>
        <body>
          <div class="container">
            <h1 id="title">Dynamic Title</h1>
            <p>This page has embedded JavaScript and CSS.</p>
            <button onclick="changeTitle()">Change Title</button>
          </div>
          
          <script>
            function changeTitle() {
              document.getElementById('title').textContent = 'Title Changed!';
            }
            document.addEventListener('DOMContentLoaded', function() {
              console.log('Page loaded with embedded scripts');
            });
          <\\/script>
        </body>
      </html>
    \`;
})();
"
`;

exports[`HTML Processing should handle HTML files with embedded scripts and styles: generated/html-processing/embedded-content/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-generated-html-processing-embedded-test-embedded.html": "test-generated-html-processing-embedded-test-embedded.html"
  },
  "assetNameToPath": {},
  "documentPrefix": "ws:///",
  "assetPrefix": "/"
}"
`;

exports[`HTML Processing should handle HTML files with embedded scripts and styles: generated/html-processing/embedded-content/test-generated-html-processing-embedded-test-embedded.html 1`] = `
"
      <!DOCTYPE html>
      <html>
        <head>
          <title>Embedded Content</title>
          <style>
            body { background-color: #f0f0f0; }
            .container { max-width: 800px; margin: 0 auto; }
            h1 { color: #333; }
          </style>
        </head>
        <body>
          <div class="container">
            <h1 id="title">Dynamic Title</h1>
            <p>This page has embedded JavaScript and CSS.</p>
            <button onclick="changeTitle()">Change Title</button>
          </div>
          
          <script>
            function changeTitle() {
              document.getElementById('title').textContent = 'Title Changed!';
            }
            document.addEventListener('DOMContentLoaded', function() {
              console.log('Page loaded with embedded scripts');
            });
          </script>
        </body>
      </html>
    "
`;

exports[`HTML Processing should handle HTML files with various content types: generated/html-processing/content-structures/complex.js 1`] = `
"(() => {
  var complex_default = '\\n      <!DOCTYPE html>\\n      <html lang="en">\\n        <head>\\n          <meta charset="UTF-8">\\n          <meta name="viewport" content="width=device-width, initial-scale=1.0">\\n          <title>Complex HTML</title>\\n        </head>\\n        <body>\\n          <header>\\n            <nav>\\n              <ul>\\n                <li><a href="#home">Home</a></li>\\n                <li><a href="#about">About</a></li>\\n              </ul>\\n            </nav>\\n          </header>\\n          <main>\\n            <section id="content">\\n              <article>\\n                <h1>Article Title</h1>\\n                <p>Some content here.</p>\\n              </article>\\n            </section>\\n          </main>\\n          <footer>\\n            <p>&copy; 2023 Test</p>\\n          </footer>\\n        </body>\\n      </html>\\n    ';
})();
"
`;

exports[`HTML Processing should handle HTML files with various content types: generated/html-processing/content-structures/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-generated-html-processing-html-test-complex.html": "test-generated-html-processing-html-test-complex.html",
    "test-generated-html-processing-html-test-simple.html": "test-generated-html-processing-html-test-simple.html"
  },
  "assetNameToPath": {},
  "documentPrefix": "ws:///",
  "assetPrefix": "/"
}"
`;

exports[`HTML Processing should handle HTML files with various content types: generated/html-processing/content-structures/simple.js 1`] = `
"(() => {
  var simple_default = "<!DOCTYPE html><html><head><title>Simple</title></head><body><h1>Hello</h1></body></html>";
})();
"
`;

exports[`HTML Processing should handle HTML files with various content types: generated/html-processing/content-structures/test-generated-html-processing-html-test-complex.html 1`] = `
"
      <!DOCTYPE html>
      <html lang="en">
        <head>
          <meta charset="UTF-8">
          <meta name="viewport" content="width=device-width, initial-scale=1.0">
          <title>Complex HTML</title>
        </head>
        <body>
          <header>
            <nav>
              <ul>
                <li><a href="#home">Home</a></li>
                <li><a href="#about">About</a></li>
              </ul>
            </nav>
          </header>
          <main>
            <section id="content">
              <article>
                <h1>Article Title</h1>
                <p>Some content here.</p>
              </article>
            </section>
          </main>
          <footer>
            <p>&copy; 2023 Test</p>
          </footer>
        </body>
      </html>
    "
`;

exports[`HTML Processing should handle HTML files with various content types: generated/html-processing/content-structures/test-generated-html-processing-html-test-simple.html 1`] = `"<!DOCTYPE html><html><head><title>Simple</title></head><body><h1>Hello</h1></body></html>"`;

exports[`HTML Processing should handle complex HTML structures and nested elements: generated/html-processing/complex-structures/complex.js 1`] = `
"(() => {
  var complex_default = '\\n      <!DOCTYPE html>\\n      <html>\\n        <head>\\n          <title>Complex Structure</title>\\n        </head>\\n        <body>\\n          <div class="wrapper">\\n            <header class="main-header">\\n              <nav class="navigation">\\n                <ul class="nav-list">\\n                  <li class="nav-item">\\n                    <a href="#section1" class="nav-link">\\n                      <span class="link-text">Section 1</span>\\n                      <span class="link-icon">\\u2192</span>\\n                    </a>\\n                    <ul class="sub-nav">\\n                      <li><a href="#subsection1">Sub 1</a></li>\\n                      <li><a href="#subsection2">Sub 2</a></li>\\n                    </ul>\\n                  </li>\\n                </ul>\\n              </nav>\\n            </header>\\n            \\n            <main class="content">\\n              <section id="section1" class="content-section">\\n                <article class="article">\\n                  <header class="article-header">\\n                    <h1 class="article-title">Complex Article</h1>\\n                    <div class="article-meta">\\n                      <time datetime="2023-01-01">January 1, 2023</time>\\n                      <address class="author">\\n                        <a rel="author" href="/author">John Doe</a>\\n                      </address>\\n                    </div>\\n                  </header>\\n                  \\n                  <div class="article-content">\\n                    <p>This is a paragraph with <strong>bold</strong> and <em>italic</em> text.</p>\\n                    <blockquote cite="https://example.com">\\n                      <p>This is a quoted paragraph.</p>\\n                      <footer>\\n                        \\u2014 <cite>Famous Author</cite>\\n                      </footer>\\n                    </blockquote>\\n                    \\n                    <figure class="image-figure">\\n                      <img src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAwIiBoZWlnaHQ9IjEwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cmVjdCB3aWR0aD0iMTAwIiBoZWlnaHQ9IjEwMCIgZmlsbD0iIzMzMzMzMyIvPjwvc3ZnPg==" alt="Test image" />\\n                      <figcaption>This is a test image caption</figcaption>\\n                    </figure>\\n                    \\n                    <details class="expandable-content">\\n                      <summary>Click to expand</summary>\\n                      <div class="hidden-content">\\n                        <p>This content was hidden.</p>\\n                        <ul>\\n                          <li>Item 1</li>\\n                          <li>Item 2</li>\\n                        </ul>\\n                      </div>\\n                    </details>\\n                  </div>\\n                </article>\\n              </section>\\n            </main>\\n            \\n            <aside class="sidebar">\\n              <section class="widget">\\n                <h3>Related Links</h3>\\n                <nav class="widget-nav">\\n                  <ul>\\n                    <li><a href="#related1">Related 1</a></li>\\n                    <li><a href="#related2">Related 2</a></li>\\n                  </ul>\\n                </nav>\\n              </section>\\n            </aside>\\n            \\n            <footer class="main-footer">\\n              <div class="footer-content">\\n                <p>&copy; 2023 Complex Structure Test</p>\\n              </div>\\n            </footer>\\n          </div>\\n        </body>\\n      </html>\\n    ';
})();
"
`;

exports[`HTML Processing should handle complex HTML structures and nested elements: generated/html-processing/complex-structures/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-generated-html-processing-complex-test-complex.html": "test-generated-html-processing-complex-test-complex.html"
  },
  "assetNameToPath": {},
  "documentPrefix": "ws:///",
  "assetPrefix": "/"
}"
`;

exports[`HTML Processing should handle complex HTML structures and nested elements: generated/html-processing/complex-structures/test-generated-html-processing-complex-test-complex.html 1`] = `
"
      <!DOCTYPE html>
      <html>
        <head>
          <title>Complex Structure</title>
        </head>
        <body>
          <div class="wrapper">
            <header class="main-header">
              <nav class="navigation">
                <ul class="nav-list">
                  <li class="nav-item">
                    <a href="#section1" class="nav-link">
                      <span class="link-text">Section 1</span>
                      <span class="link-icon">→</span>
                    </a>
                    <ul class="sub-nav">
                      <li><a href="#subsection1">Sub 1</a></li>
                      <li><a href="#subsection2">Sub 2</a></li>
                    </ul>
                  </li>
                </ul>
              </nav>
            </header>
            
            <main class="content">
              <section id="section1" class="content-section">
                <article class="article">
                  <header class="article-header">
                    <h1 class="article-title">Complex Article</h1>
                    <div class="article-meta">
                      <time datetime="2023-01-01">January 1, 2023</time>
                      <address class="author">
                        <a rel="author" href="/author">John Doe</a>
                      </address>
                    </div>
                  </header>
                  
                  <div class="article-content">
                    <p>This is a paragraph with <strong>bold</strong> and <em>italic</em> text.</p>
                    <blockquote cite="https://example.com">
                      <p>This is a quoted paragraph.</p>
                      <footer>
                        — <cite>Famous Author</cite>
                      </footer>
                    </blockquote>
                    
                    <figure class="image-figure">
                      <img src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAwIiBoZWlnaHQ9IjEwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cmVjdCB3aWR0aD0iMTAwIiBoZWlnaHQ9IjEwMCIgZmlsbD0iIzMzMzMzMyIvPjwvc3ZnPg==" alt="Test image" />
                      <figcaption>This is a test image caption</figcaption>
                    </figure>
                    
                    <details class="expandable-content">
                      <summary>Click to expand</summary>
                      <div class="hidden-content">
                        <p>This content was hidden.</p>
                        <ul>
                          <li>Item 1</li>
                          <li>Item 2</li>
                        </ul>
                      </div>
                    </details>
                  </div>
                </article>
              </section>
            </main>
            
            <aside class="sidebar">
              <section class="widget">
                <h3>Related Links</h3>
                <nav class="widget-nav">
                  <ul>
                    <li><a href="#related1">Related 1</a></li>
                    <li><a href="#related2">Related 2</a></li>
                  </ul>
                </nav>
              </section>
            </aside>
            
            <footer class="main-footer">
              <div class="footer-content">
                <p>&copy; 2023 Complex Structure Test</p>
              </div>
            </footer>
          </div>
        </body>
      </html>
    "
`;

exports[`HTML Processing should handle mml: imports of HTML files: generated/html-processing/mml-html-imports/c/d.js 1`] = `
"(() => {
  var d_default = "<html>\\n  <body>\\n    <p>I'm html!</p>\\n  </body>\\n</html>\\n";
})();
"
`;

exports[`HTML Processing should handle mml: imports of HTML files: generated/html-processing/mml-html-imports/html-importer.js 1`] = `
"(() => {
  var d_default = "ws:///test-src-c-d.html";
  function html_importer_default() {
    return d_default;
  }
})();
"
`;

exports[`HTML Processing should handle mml: imports of HTML files: generated/html-processing/mml-html-imports/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-src-c-d.html": "test-src-c-d.html",
    "test-src-html-importer.html": "test-src-html-importer.html"
  },
  "assetNameToPath": {},
  "documentPrefix": "ws:///",
  "assetPrefix": "/"
}"
`;

exports[`HTML Processing should handle mml: imports of HTML files: generated/html-processing/mml-html-imports/test-src-c-d.html 1`] = `
"<html>
  <body>
    <p>I'm html!</p>
  </body>
</html>
"
`;

exports[`HTML Processing should handle mml: imports of HTML files: generated/html-processing/mml-html-imports/test-src-html-importer.html 1`] = `
"<body></body><script>(() => {
  var d_default = "ws:///test-src-c-d.html";
  function html_importer_default() {
    return d_default;
  }
})();
</script>"
`;

exports[`HTML Processing should process the existing c/d.html file: generated/html-processing/subdirectory-html/d.js 1`] = `
"(() => {
  var d_default = "<html>\\n  <body>\\n    <p>I'm html!</p>\\n  </body>\\n</html>\\n";
})();
"
`;

exports[`HTML Processing should process the existing c/d.html file: generated/html-processing/subdirectory-html/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-src-c-d.html": "test-src-c-d.html"
  },
  "assetNameToPath": {},
  "documentPrefix": "ws:///",
  "assetPrefix": "/"
}"
`;

exports[`HTML Processing should process the existing c/d.html file: generated/html-processing/subdirectory-html/test-src-c-d.html 1`] = `
"<html>
  <body>
    <p>I'm html!</p>
  </body>
</html>
"
`;
